language/move-prover/bytecode/src/borrow_analysis.rs

```diff
@@ -254,90 +203,170 @@ impl BorrowInfo {
     }
 
     fn add_edge(&mut self, parent: BorrowNode, child: BorrowNode, weight: BorrowEdge) -> bool {
-        if self.unchecked_nodes.contains(&parent) {
-            // If the parent node is unchecked, so is the child node.
-            self.unchecked_nodes.insert(child.clone());
-        }
-        match self
-            .borrowed_by
-            .entry(parent.clone())
-            .or_default()
-            .entry(child.clone())
+        self.borrowed_by
+            .entry(parent)
             .or_default()
-        {
-            EdgeDomain::Top => false,
-            EdgeDomain::EdgeSet(edges) => match weight {
-                BorrowEdge::Weak => {
-                    self.borrowed_by
-                        .entry(parent)
-                        .or_default()
-                        .insert(child, EdgeDomain::Top);
-                    true
-                }
-                BorrowEdge::Strong(edge) => edges.insert(edge),
-            },
-        }
+            .insert((child, weight))
     }
 
     fn consolidate(&mut self) {
-        for (src, outgoing) in &self.borrowed_by {
-            for (dst, edges) in outgoing {
+        for (src, outgoing) in self.borrowed_by.iter() {
+            for (dst, edge) in outgoing.iter() {
                 self.borrows_from
                     .entry(dst.clone())
                     .or_default()
-                    .insert(src.clone(), edges.clone());
+                    .insert((src.clone(), edge.clone()));
+            }
+        }
+    }
+
+    /// Collect those leaves which are returned and summarize them in a hyper edge.
+    /// Each of those leaves has a path `out_mut -> ref1 .. -> refn -> in_mut`.
```

Comment:
Have you swapped in_mut and out_mut in the comment?  The hyper-edge is from the input mutable reference to the output mutable reference as far as I understand.